modules/loader/source/relocations.cpp

```diff
@@ -645,88 +650,261 @@ bool resolveAArch64(const loader::Relocation &r, loader::ElfMap &map,
   }
   return true;
 }
+
+// Sets the immediate field of an I-type instruction [31:20]
+void writeITypeImm(uint32_t imm, uint8_t *insn_addr) {
+  uint32_t insn;
+  cargo::read_little_endian(&insn, insn_addr);
+  insn = setBitRange(insn, imm, 20, 12);
+  cargo::write_little_endian(insn, insn_addr);
+}
+
+// Sets the immediate field of an U-type instruction [31:12]
+void writeUTypeImm(uint32_t imm, uint8_t *insn_addr) {
+  uint32_t insn;
+  cargo::read_little_endian(&insn, insn_addr);
+  insn = setBitRange(insn, imm, 12, 20);
+  cargo::write_little_endian(insn, insn_addr);
+}
+
+uint32_t getLo12(uint64_t addr) {
+  return static_cast<uint32_t>(getBitRange(addr, 0, 12));
+}
+
+std::optional<uint32_t> getHi20(int64_t addr) {
+  auto signExtend32 = [](int64_t val) {
+    return static_cast<int64_t>(val << 32) >> 32;
+  };
+  addr += 0x800;
+  if (addr != signExtend32(addr)) {
+    (void)fprintf(
+        stderr,
+        "Value %" PRId64 " is out of the range of a 32-bit immediate.\n", addr);
+    return std::nullopt;
+  }
+  return static_cast<uint32_t>(getBitRange(addr, 12, 20));
+}
+
+// assumes little-endian RISCV
+bool resolveRISCV(const loader::Relocation &r, loader::ElfMap &map,
+                  const std::vector<loader::Relocation> &relocations) {
+  const auto relocation_data = decomposeRelocation<uint64_t>(r, map);
+  if (!relocation_data) {
+    return false;
+  }
+  auto [relocation_address, relocation_target_address, symbol_target_address] =
+      *relocation_data;
+
+  // A PC-relative offset (S + A - P)
```

Comment:
I'm not sure what the comment refers to , is it just to explain L699 or all the relative offsets later on?

Reply:
Both, I guess. It's to explain the variable `relative_value` and everywhere it's used

Reply:
not sure what you are referring to exactly, but if you mean a PC-relative value in relocations such as `PCREL_LO12_I` which are just `S - P`, you're right, but that's because the addend `A` must be zero for those relocations. Other PC-relative values like `PCREL_HI20` must include the addend `A`.

Since `symbol_target_address` is `S + A` I think this documentation is consistent.